metrics/prometheus/registry.go

```diff
@@ -35,97 +34,84 @@ import (
-	svm map[string]*prom.SummaryVec   // prom.SummaryVec
+	r    prom.Registerer // for convenience of testing
+	vecs sync.Map
 }
 
-func (p *promMetricRegistry) Counter(m *metrics.MetricId) metrics.CounterMetric {
-	p.mtx.RLock()
-	vec, ok := p.cvm[m.Name]
-	p.mtx.RUnlock()
+func (p *promMetricRegistry) getOrComputeVec(key string, supplier func() interface{}) interface{} {
+	v, ok := p.vecs.Load(key)
 	if !ok {
-		p.mtx.Lock()
-		vec = promauto.NewCounterVec(prom.CounterOpts{
+		v, ok = p.vecs.LoadOrStore(key, supplier())
+		if !ok {
+			p.r.MustRegister(v.(prom.Collector)) // only registe collector which stored success
+		}
+	}
+	return v
+}
+
+func (p *promMetricRegistry) Counter(m *metrics.MetricId) metrics.CounterMetric {
+	vec := p.getOrComputeVec(m.Name, func() interface{} {
+		return prom.NewCounterVec(prom.CounterOpts{
 			Name: m.Name,
 			Help: m.Desc,
 		}, m.TagKeys())
-		p.cvm[m.Name] = vec
-		p.mtx.Unlock()
-	}
-	c := vec.With(m.Tags)
-	return &counter{pc: c}
+	}).(*prom.CounterVec)
+	return vec.With(m.Tags)
 }
 
 func (p *promMetricRegistry) Gauge(m *metrics.MetricId) metrics.GaugeMetric {
-	p.mtx.RLock()
-	vec, ok := p.gvm[m.Name]
-	p.mtx.RUnlock()
-	if !ok {
-		p.mtx.Lock()
-		vec = promauto.NewGaugeVec(prom.GaugeOpts{
+	vec := p.getOrComputeVec(m.Name, func() interface{} {
+		return prom.NewGaugeVec(prom.GaugeOpts{
 			Name: m.Name,
 			Help: m.Desc,
 		}, m.TagKeys())
-		p.gvm[m.Name] = vec
-		p.mtx.Unlock()
-	}
-	g := vec.With(m.Tags)
-	return &gauge{pg: g}
+	}).(*prom.GaugeVec)
+	return vec.With(m.Tags)
 }
 
-func (p *promMetricRegistry) Histogram(m *metrics.MetricId) metrics.HistogramMetric {
-	p.mtx.RLock()
-	vec, ok := p.hvm[m.Name]
-	p.mtx.RUnlock()
-	if !ok {
-		p.mtx.Lock()
-		vec = promauto.NewHistogramVec(prom.HistogramOpts{
+func (p *promMetricRegistry) Histogram(m *metrics.MetricId) metrics.ObservableMetric {
+	vec := p.getOrComputeVec(m.Name, func() interface{} {
+		return prom.NewHistogramVec(prom.HistogramOpts{
 			Name: m.Name,
 			Help: m.Desc,
 		}, m.TagKeys())
-		p.hvm[m.Name] = vec
-		p.mtx.Unlock()
-	}
-	h := vec.With(m.Tags)
-	return &histogram{ph: h.(prom.Histogram)}
+	}).(*prom.HistogramVec)
+	return vec.With(m.Tags)
 }
 
-func (p *promMetricRegistry) Summary(m *metrics.MetricId) metrics.SummaryMetric {
-	p.mtx.RLock()
-	vec, ok := p.svm[m.Name]
-	p.mtx.RUnlock()
-	if !ok {
-		p.mtx.Lock()
-		vec = promauto.NewSummaryVec(prom.SummaryOpts{
+func (p *promMetricRegistry) Summary(m *metrics.MetricId) metrics.ObservableMetric {
+	vec := p.getOrComputeVec(m.Name, func() interface{} {
+		return prom.NewSummaryVec(prom.SummaryOpts{
 			Name: m.Name,
 			Help: m.Desc,
 		}, m.TagKeys())
-		p.svm[m.Name] = vec
-		p.mtx.Unlock()
-	}
-	s := vec.With(m.Tags)
-	return &summary{ps: s.(prom.Summary)}
+	}).(*prom.SummaryVec)
+	return vec.With(m.Tags)
 }
 
-func (p *promMetricRegistry) Export() {
+func (p *promMetricRegistry) Rt(m *metrics.MetricId) metrics.ObservableMetric {
```

Comment:
I think it is more appropriate to call it AggregateRT, and RT should be a set of  global metrics without using time window.

Reply:
Maybe I should change this to a normal rt implementation instead of an aggregate rt implementation, because the aggregate rt implementation is quite different from this